lib/matplotlib/backend_bases.py

```diff
@@ -2561,6 +2612,128 @@ class NonGuiException(Exception):
     pass
 
 
+class WindowEvent(object):
+    def __init__(self, name, window):
+        self.name = name
+        self.window = window
+
+
+class WindowBase(cbook.EventEmitter):
+    """The base class to show a window on screen.
+
+    Parameters
+    ----------
+    title : str
+        The title of the window.
+    """
+
+    def __init__(self, title):
+        cbook.EventEmitter.__init__(self)
+
+    def show(self):
+        """
+        For GUI backends, show the figure window and redraw.
+        For non-GUI backends, raise an exception to be caught
+        by :meth:`~matplotlib.figure.Figure.show`, for an
+        optional warning.
+        """
+        raise NonGuiException()
+
+    def destroy(self):
+        """Destroys the window"""
+        pass
+
+    def set_fullscreen(self, fullscreen):
+        """Whether to show the window fullscreen or not, GUI only.
+
+        Parameters
+        ----------
+        fullscreen : bool
+            True for yes, False for no.
+        """
+        pass
+
+    def set_default_size(self, width, height):
+        """Sets the default size of the window, defaults to a simple resize.
+
+        Parameters
+        ----------
+        width : int
+            The default width (in pixels) of the window.
+        height : int
+            The default height (in pixels) of the window.
+        """
+        self.resize(width, height)
+
+    def resize(self, width, height):
+        """"For gui backends, resizes the window.
+
+        Parameters
+        ----------
+        width : int
+            The new width (in pixels) for the window.
+        height : int
+            The new height (in pixels) for the window.
+        """
+        pass
+
+    def get_window_title(self):
+        """
+        Get the title text of the window containing the figure.
+        Return None for non-GUI backends (e.g., a PS backend).
+
+        Returns
+        -------
+        str : The window's title.
+        """
+        return 'image'
+
+    def set_window_title(self, title):
+        """
+        Set the title text of the window containing the figure.  Note that
+        this has no effect for non-GUI backends (e.g., a PS backend).
+
+        Parameters
+        ----------
+        title : str
+            The title of the window.
+        """
+        pass
+
+    def add_element_to_window(self, element, expand, fill, pad, place):
```

Comment:
@tacaswell Pedantically speaking, the mpl-backend defined GUI widgets exist as native GUI widgets as they all extend from a base native-GUI class.

```python
class NavigationToolbar2GTK3(NavigationToolbar2, Gtk.Toolbar):
```

so hectically speaking one can pass any native widget to this function, though we would expect only ``Toolbar`` ``NavigationToolbar2``, ``Canvas`` or ``Statusbar`` (I believe what fariza called ``Message``)

If I understand your question correctly @tacaswell you feel concerned that people will assume that they can add an mpl widget from widgets.py to this and that if we don't allow those widgets then we should not use ``widget`` in the title.  Your question made me think whether we could add in other widgets with something like...
```
if isinstance(widget, matplotlib.widgets.AxesWidget):
  widget = widget.canvas
```
until I remembered that multiple widgets will share the same canvas, and we don't want to try adding the same canvas multiple times to the same window, and I don't think we want to promote each widget having its own canvas, so...